### ISSUE TITLE
Custom option price not reflecting catalog price rule

### ISSUE THREAD

**Issue by johncarlson21** [magento/magento2#21879]
Products with custom option and price are incorrect.
Example: product price = 7.00; custom option = 0.50
30% discount applied to product.

### Preconditions (*)

1. Magento 2.3.0
2. PHP 7.2

### Steps to reproduce (*)

1. Create Simple Product with a price of 7.00 and add a custom option with a price with a price of 0.50
2. Create a 30% catalog price rule for this product and apply rules
3. Go to Frontend product choose the option.

### Expected result (*)

1. product price should be 30% off 7.50
2. 7.50 - (7.50 * .3) = 5.25

### Actual result (*)

1. 7 - (7 * .3) + .5 = 5.40
2. 30% discount is applied to the base price but not to the custom option price.

**Comment by magento-engcom-team** [comment]
Hi @johncarlson21. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me 2.3-develop instance` - upcoming 2.3.x release

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@johncarlson21 do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by johncarlson21** [comment]
@magento-engcom-team give me 2.3-develop instance - upcoming 2.3.x release

**Comment by magento-engcom-team** [comment]
Hi @johncarlson21. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @johncarlson21, here is your [Magento instance](https://i-21879-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-21879-2-3-develop.instances.magento-community.engineering/admin](https://i-21879-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by johncarlson21** [comment]
Confirmed in the vanilla magento instance that my issue exists. See screenshots.

<img width="1370" alt="Screen Shot 2019-03-21 at 11 46 34 AM" src="https://user-images.githubusercontent.com/25893/54769400-1be7cb80-4bcf-11e9-9dbc-46c61643ac98.png">
<img width="1373" alt="Screen Shot 2019-03-21 at 11 47 12 AM" src="https://user-images.githubusercontent.com/25893/54769409-21451600-4bcf-11e9-99ab-0c863e5dc579.png">

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

-  _Next steps are available in case you are a member of **Community Maintainers**._

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-backlog-nazar 
 Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-98841`, `MAGETWO-98842` were created

**Issue Available**: @engcom-backlog-nazar, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by johncarlson21** [comment]
Can I get an update on this please? we are looking at launching an updated site very soon and this is a pertinent functionality as all our products use this.

**Comment by johncarlson21** [comment]
Please provide some feedback on this issue as this is a needed functionality.

**Comment by parekhreena** [comment]
Hi @magento-engcom-team , 

Is there any update on this please ?

**Comment by dhorytskyi** [comment]
@johncarlson21 Hi! As I know, currently it's behaviour by design.

**Comment by johncarlson21** [comment]
this is currently how it is, but in Magento 1.9.x etc.. and previous versions, it is not like this.
the thing is, when you are working with a product, the promotions should also affect the product option pricing. this needs to be standard functionality or people will not understand why their pricing doesn't match up with how it should be.

**Comment by johncarlson21** [comment]
@Surabhi-Cedcoss has anyone gotten to look at this yet?

**Comment by johncarlson21** [comment]
any word?

**Comment by p-bystritsky** [comment]
I'm already working on the same issue https://github.com/magento/magento2/pull/22917

**Comment by johncarlson21** [comment]
@p-bystritsky  I looked over at your link.. could you tell me the status of this? is it in testing? is it done and ready for the next update to code?

**Comment by Techytt** [comment]
any latest update on this issue?

**Comment by johncarlson21** [comment]
@Techytt this will be included in the 2.4 update I believe.. I took the above changes and added them as an extension to mine for now.

**Comment by gsbhati2008** [comment]
@johncarlson21 can you please send me extension that you created for now. 
Thanks

**Comment by johncarlson21** [comment]
I will try to get this tonight or tomorrow. 

John

**Comment by gsbhati2008** [comment]
@johncarlson21 Thanks

**Comment by johncarlson21** [comment]
@gsbhati2009 here it is.. just install like a normal extension and enable it.
[Surething_Calculateops.zip](https://github.com/magento/magento2/files/4275826/Surething_Calculateops.zip)

**Comment by gsbhati2008** [comment]
Thanks @johncarlson21 for extension .
I installed extension and is having this error:
Fatal error: Access level to Surething\Calculateops\CatalogRule\Model\Rule::$conditionsToCollectionApplier must be protected (as in class Magento\CatalogRule\Model\Rule) or weaker in /home/mage/public_html/app/code/Surething/Calculateops/CatalogRule/Model/Rule.php on line 53
I am using Magento ver. 2.2.8. Please help me fix this issue.

**Comment by johncarlson21** [comment]
I've created this for magento 2.3.x but I think the fix is this:
in file: app/code/Surething/Calculateops/CatalogRule/Model/Rule.php
line 169
private $conditionsToCollectionApplier;
change to
protected $conditionsToCollectionApplier;

there are other ones below that might throw an error 
private $websitesMap; on line 174
private $ruleResourceModel; on line 179
they might need to be changed to protected also.

**Comment by gsbhati2008** [comment]
@johncarlson21 thanks for corrections, its working.
Thanks a lot man!!!

**Comment by gsbhati2008** [comment]
@johncarlson21 prices are not calculating properly for multi curreny. Can you please help me in this.
I am attaching images of issue.
![image](https://user-images.githubusercontent.com/61616076/76697372-77527980-66bc-11ea-9166-67197b1857cf.png)
here you see price 421.80
After selecting 
![image](https://user-images.githubusercontent.com/61616076/76697395-ac5ecc00-66bc-11ea-9e11-1e65a1978935.png)
Price is $ 434.01
Euro is base Currency and Price for this option is 380 Euro.
Currency rate in admin is 1.11 (euro to usd) .
So the final price after selection must be 380 x 1.11=$421.80.
If I disable extension prices are fine. 
It would be great if you help in this.
Thanks

**Comment by gsbhati2008** [comment]
@johncarlson21 I debug little more and what I found is that.
I have product price 280 Euro and option price 100 Euro.
So total is 380 Euro. Currency rate in admin from Euro to USD is 1.11.
So value must be (280+100)x1.11= $421.80.
However its calculating like ((100 x 1.11)+280 )x 1.11.
So it becomes like (111+280) x 1.11
(390) x 1.11= $434.01.

So in short its first multiplying currency rate with options, then it adds that with product price.
Then finally Multiplies again with currency rate.
In short custom options price is multiplied two times with currency rate.
@johncarlson21  can you please help me in fixing this.

Thanks

**Comment by gsbhati2008** [comment]
HI @johncarlson21 , I found the fix.
This line return $this->priceCurrency->convertAndRound($finalOptionPrice); 
in Surething/Calculateops/Catalog/Pricing/Price/CalculateCustomOptionCatalogRule.php line 80 can be changed to
return $finalOptionPrice;

Thanks

**Comment by johncarlson21** [comment]
Great!
I wonder if this is a problem with you using 2.2.x and not 2.3.x
the code was written to fix issues in 2.3.x
I didn't write the code, I just put it into an extension for now :) also I don't have a store setup to do the currency stuff and test at the moment.

**Comment by codedge** [comment]
Thanks for keeping track of this @johncarlson21 
I can confirm the exact bug on 2.3.4. I am going to try the [extension you provided ](https://github.com/magento/magento2/issues/21879#issuecomment-593425596). Hopefully this is going to be fixed in the next version.

**Comment by udovicic** [comment]
I must agree with @dhorytskyi , this is behavior by design. At least, that you will get from EE support.

**Comment by magento-engcom-team** [comment]
The report related to technology or functionality that was removed(or not supported) from the Magento code-base.
Or cannot be fixed and delivered in the scope of this public repository.

Examples:
- Removed payment methods. [Magento Commerce 2.4.0 Release Notes](https://devdocs.magento.com/guides/v2.4/release-notes/release-notes-2-4-0-commerce.html), [Deprecation of Magento core payment integrations](https://community.magento.com/t5/Magento-DevBlog/Deprecation-of-Magento-core-payment-integrations/ba-p/426445). All questions, issue reports, and fix for them should be addressed to the corresponding extension owners(support) on the [Magento Market place](https://marketplace.magento.com/extensions.html) page
- Mysql search. Replaced with Elasticsearch 
- Not [Supported browsers](https://devdocs.magento.com/guides/v2.4/install-gde/system-requirements.html#supported-browsers)
- Not supported [Operating systems](https://devdocs.magento.com/guides/v2.4/install-gde/system-requirements.html#operating-systems-linux-x86-64)
- Web Wizard install
- Not a bug 
- etc

**Comment by magento-engcom-team** [comment]
The report related to technology or functionality that was removed(or not supported) from the Magento code-base.
Or cannot be fixed and delivered in the scope of this public repository.

Examples:
- Removed payment methods. [Magento Commerce 2.4.0 Release Notes](https://devdocs.magento.com/guides/v2.4/release-notes/release-notes-2-4-0-commerce.html), [Deprecation of Magento core payment integrations](https://community.magento.com/t5/Magento-DevBlog/Deprecation-of-Magento-core-payment-integrations/ba-p/426445). All questions, issue reports, and fix for them should be addressed to the corresponding extension owners(support) on the [Magento Market place](https://marketplace.magento.com/extensions.html) page
- Mysql search. Replaced with Elasticsearch 
- Not [Supported browsers](https://devdocs.magento.com/guides/v2.4/install-gde/system-requirements.html#supported-browsers)
- Not supported [Operating systems](https://devdocs.magento.com/guides/v2.4/install-gde/system-requirements.html#operating-systems-linux-x86-64)
- Web Wizard install
- Not a bug 
- etc

**Comment by maghamed** [comment]
@magento give me 2.4-develop instance - upcoming 2.4.x release

**Comment by maghamed** [comment]
@magento give me 2.4-develop instance - upcoming 2.4.x release

**Comment by maghamed** [comment]
@magento give me 2.4-develop instance - upcoming 2.4.x release

**Comment by maghamed** [comment]
@magento give me 2.4-develop instance - upcoming 2.4.x release

**Comment by maghamed** [comment]
@magento give me 2.4-develop instance - upcoming 2.4.x release

**Comment by maghamed** [comment]
@magento give me 2.4-develop instance - upcoming 2.4.x release
